Fixes #15

maedot-d opened the pull request:
Clean sockets stuck at CLOSE_WAIT/FIN_WAIT2 state,
and add SO_KEEPALIVE option to clean sockets stuck at ESTABLISHED state
	modified:   src/tcp-intercept.cxx

maedotd-afk committed:
Changes to be committed: Clean sockets stuck at CLOSE_WAIT/FIN_WAIT2 state,
  and add SO_KEEPALIVE option to clean sockets stuck at ESTABLISHED state
	modified:   src/tcp-intercept.cxx

nielslaukens commented:
In what order have you implemented and tested these changes? I believe that only adding the SO_KEEPALIVE will also cleanup connections in FIN_WAIT2 and CLOSE_WAIT automatically by the kernel. If so, this would remove a big part of the code, which would make tcp-intercept faster and more reliable.

Also: You need to update the documentation (Readme.md) to note that the kernel needs to have Keepalives enabled as well for this to work (net.ipv4.tcp_keepalive_time, net.ipv4.tcp_keepalive_intvl, net.ipv4.tcp_keepalive_probes).

maedot-d commented:
Hi,

I fixed the close_wait and fin_wait2 issues first, as they were the main 
problem. And then the established ones.

I agree with you in principle. So I will test only the keepalive fix and 
let you know.

Maedot

-----Original-Nachricht-----
Von: Niels Laukens <notifications@github.com
<mailto:notifications@github.com> >
Betreff: Re: [VRT-onderzoek-en-innovatie/tcp-intercept] Fixes #15 (#16)
Datum: 19.01.2017, 12:58 Uhr
An: VRT-onderzoek-en-innovatie/tcp-intercept <
tcp-intercept@noreply.github.com <mailto:tcp-intercept@noreply.github.com> 
>
CC: maedot-d <maedot@t-online.de <mailto:maedot@t-online.de> >, Author <
author@noreply.github.com <mailto:author@noreply.github.com> >

In what order have you implemented and tested these changes? I believe that 
only adding the SO_KEEPALIVE will also cleanup connections in FIN_WAIT2 and 
CLOSE_WAIT automatically by the kernel. If so, this would remove a big part 
of the code, which would make tcp-intercept faster and more reliable.

Also: You need to update the documentation (Readme.md) to note that the 
kernel needs to have Keepalives enabled as well for this to work 
(net.ipv4.tcp_keepalive_time, net.ipv4.tcp_keepalive_intvl, 
net.ipv4.tcp_keepalive_probes).

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub
<https://github.com/VRT-onderzoek-en-innovatie/tcp-intercept/pull/16#issuecomment-273758404> 
, or mute the thread
<https://github.com/notifications/unsubscribe-auth/AHT7TRyBc_wNNNFJoLmcQrzHDN67UjRvks5rT0_KgaJpZM4Ln8ZX> 
.

maedot-d commented:
Ok, 

I made tests and it seems the keepalive is sufficient. But one problem with that is, I am able to set the keepalive option only on the s_server socket not s_client socket, which would hang in the FIN_WAIT2 state otherwise.
Is it not possible to add this option for a client socket? or am I missing something?

maedot-d committed:
Update README.md

maedot-d committed:
Update README.md

maedot-d committed:
Update README.md

maedot-d committed:
Update README.md

nielslaukens commented:
What do you mean with being unable to set it? Does it return an error (if so: which)?

Your changes to the README look good!

maedot-d commented:
It does not return any errors, but does not activate the keepalive on the socket either. The netstat output for the socket shows timer state "off (0.00/0/0)", while for the server socket shows "keepalive (589.91/0/0)". Without keepalive on the client socket, it is necessary to keep the patch as a whole, to avoid resource leak.

cpaasch commented:
I haven't followed through all of this thread here, but just to clarify:

Setting SO_KEEPALIVE on an MPTCP-socket will not set the keepalive timer on the subflows, but rather on the MPTCP-socket. When the timer expires, it sends a keepalive on all subflows.

Thus, netstat will not show the keepalive timer on the subflows.

nielslaukens commented on src/tcp-intercept.cxx:
```diff
@@ -473,6 +505,9 @@ int main(int argc, char* argv[]) {
 		s_listen.setsockopt(SOL_IP, IP_TRANSPARENT, &value, sizeof(value));
+		s_listen.setsockopt(IPPROTO_TCP, TCP_NODELAY, (char *) &value, sizeof(value));
+		//Take care of socks hanging in Established state
+		s_listen.setsockopt(SOL_SOCKET,SO_KEEPALIVE, &value, sizeof(value));
```
There is no need to set the listening socket with Keepalive. This socket will only receive new connections, so will never get out of the LISTENING state.

You should move this (and probably the NO_DELAY as well) to around here: https://github.com/maedot-d/tcp-intercept/blob/821a9f89a06c893c1648d547d2080e4b21a3e7ce/src/tcp-intercept.cxx#L271

maedot-d replied:
- Activate/deactivate TCP_NODELAY:
  This could be done via sysctl interface:
       net.ipv4.tcp_low_latency = 0/1
- Setting the keepalive option on the listening socket:
That was not my original idea either. After a day of unsuccessful attempts to set it only on the s_server socket, I tried (just out of curiosity) to set it on the listening socket and to my surprise it worked. I am still puzzled as to why!  I just made some tests and seems setting the option in the s_socket is redundant and could be removed.

maedot-d commented:
That is good to know! 
But I have not tested such scenario yet (did not see the need for). I am testing on regular TCP connections, which, I assume, would translate to MPTCP too.

nielslaukens reviewed (changes requested):
Also: it would be nice if both SO_KEEPALIVE and TCP_NODELAY can be activated/deactivated with command line options.

nielslaukens commented:
Thanks @cpaasch for the very valuable input!

maedotd-afk committed:
Changes to be committed: Remove unnecessary lines of code.
	modified:   src/tcp-intercept.cxx

maedotd-afk committed:
Merge branch 'fixes-#15' of https://github.com/maedot-d/tcp-intercept into fixes-#15

maedotd-afk committed:
Changes to be committed: Added notes for TCP_NODELAY
	modified:   README.md

maedot-d committed:
Update README.md

maedotd-afk committed:
Changes to be committed: Resource leak fix with keepalive only.
	modified:   src/tcp-intercept.cxx

maedotd-afk committed:
Merge branch 'fixes-#15' of https://github.com/maedot-d/tcp-intercept into fixes-#15

maedot-d commented:
Hi Niels,
It is now fixed. The "KEEPALIVE only" option works. I have updated the branch accordingly.

nielslaukens reviewed (changes requested):
Do you plan to make these optional? i.e. adding a command line flag to enable/disable SO_KEEPALIVE and/or TCP_NODELAY?

maedot-d commented on src/tcp-intercept.cxx:
```diff
@@ -243,6 +244,14 @@ static void listening_socket_ready_for_read(EV_P_ ev_io *w, int revents) {
 		new_con->s_client = s_listen->accept(&client_addr);
 
+		int val = 1;
+		//We do not want to buffer small packets, which could increase latency/gitter
```
I think that would be the best option, but at the moment I do not know how  .... need to read about it.

maedot-d commented on README.md:
```diff
@@ -46,6 +46,36 @@ make
+effect on latency/jitter sensitive applications (real time apps). Thus, enabling
+the tcp option TCP_NODELAY instructs the kernel to send data as soon as they are
+sent by the application to the kernel.
+To benefit from this feature you need to enable the corresponding kernel parameter:
```
Oh, that is a big mix up!!! Sorry my bad. I will correct it.

maedotd-afk committed:
Changes to be committed: Corrected TCP_NODELAY description,
   made keepalive and TCP_NODELAY features optional.
	modified:   README.md
	modified:   src/tcp-intercept.cxx

nielslaukens commented on src/tcp-intercept.cxx:
```diff
@@ -381,6 +394,8 @@ int main(int argc, char* argv[]) {
 					"Options:\n"
 					"  -h --help                       Displays this help message and exits\n"
 					"  -V --version                    Displays the version and exits\n"
+					"  -k --SO_KEEPALIVE               Enable keepalive on the sockets\n"
```
--keepalive (see line 377 above)

nielslaukens commented on src/tcp-intercept.cxx:
```diff
@@ -381,6 +394,8 @@ int main(int argc, char* argv[]) {
 					"  -h --help                       Displays this help message and exits\n"
 					"  -V --version                    Displays the version and exits\n"
+					"  -k --SO_KEEPALIVE               Enable keepalive on the sockets\n"
+					"  -n --TCP_NODELAY                Disable Nagel Algorithm on the sockets\n"
```
--tcp_nodelay (see line 378)

nielslaukens commented on src/tcp-intercept.cxx:
```diff
@@ -273,10 +279,15 @@ static void listening_socket_ready_for_read(EV_P_ ev_io *w, int revents) {
+			new_con->s_server.setsockopt(SOL_SOCKET, SO_KEEPALIVE, &val, sizeof(val));
+			//LogInfo(_("%1$s: Enabling keepalive on s_server"), new_con->id.c_str());		
+		}
+		if(NODELAY){
```
Very minor remark:
In the client section, you first do nodelay, then keepalive; Here, you do it the other way around... It would look better if these orders where the same. (But technically, this does not matter)

nielslaukens commented on src/tcp-intercept.cxx:
```diff
@@ -22,6 +22,8 @@ std::string logfilename;
 
 static const int MAX_CONN_BACKLOG = 32;
+bool KEEPALIVE = false;
+bool NODELAY = false;
```
I prefer lowercase variables (vs uppercase constants), but this is a minor issue.
Also: group them with the 2 vars below (line 28,29) instead of with the constants above (line 24)

nielslaukens reviewed (changes requested):
Looks very good. Only the option names need to match between the help text and the getopt() definition

maedotd-afk committed:
Changes to be committed: Cosmetic changes.
	modified:   src/tcp-intercept.cxx
